# Travail pratique d'optimisation Combinatoire

* Objectifs :

ce travail pratique s'inscrit dans une série de mise en oeuvre des différentes heuristiques étudiées, dans le but de résoudre le problème de sac à dos qui est un problème classique en optimisation combinatoire.

* Avant-propos

On sait que les variables binaires servent à représenter si un événement se réalise ou non. 

Et donc de façon simple, dans la théorie des graphes, cela modélise plutot les chemins que l'on décide de séléctionner ou pas.

Aussi, la recherche opérationnelle est une discipline à cheval entre les maths, l'informatique et l'économie, l'optimisation combinatoire quant à elle est une branche qui proprose des heuristiques (c'est à dire des méthodes de résolutions intuitives, qui servent à la découverte des solutions) en se basant sur ce principe de 1 et de 0 à l'aide des variables binaires.

Dans le cadre du cours de première année de master à l'Université de Rouen, nous avons pris plaisir à découvrir les heuristiques les plus populaires en utilisant le problème de sac à dos.

Constat : nous avons compris que tout problème peut se résoudre à l'aide d'une heuristique et, qu'il est donc possible de résoudre <plusieurs applications de la recherche opérationnelle que ce soit, celui du "voyageur de commerce", "celui du postier chinois" en passant par "le problème de transport" et "d'affectation">,
à l'aide d'un bon nombre d'heuristiques ou de méthodologie existante.

Nous osons donc dire que <toutes heuristiques mènent à Rome>, mais dans notre cas nous verrons plus loin que en Optimisation Combinatoire Rome n'est pas unique, car il peut exister plusieurs solutions candidates à l'éléction de solution optimale et ce, en fonction des méthodes tel que nous le démontre Branche and Bound.

* Le problème du sac à dos

il s'agit du problème de chargement, étant donné d'une part une collection d'objets caractérisés chacun par un volume et une valeur, et d'autre part un sac de volume limité, on charche à svoir quels objets faut-il charger dans le sac pour maximiiser la valeur totale du chargement.

* Importation des Librairies

In [2]:
from methods.knapsack_solver import run_knapsack
from methods.knapsack_greedy_solver import run_greedy
from methods.knapsack_recursivity import run_recursivity
from methods.knapsack_recursivity_memo import run_recursivity_memo
from methods.knapsack_dynamic_programmation import run_bottum_up
from methods.knapsack_BranchAndBound import KPBB
from methods.knapsack_recuit_simule import run_annealing
from methods.knapsack_genetique import kp_genetic
#from methods.knapsack_tabou_search import run_ts

from methods.others_functions import get_datas_in_an_other_format, get_knapsack_datas

In [3]:
import numpy as np

### 0 - Définition des données simples

In [141]:
cap = 10
values = [40, 50, 100, 95, 30]
weights = [2, 3, 1, 5, 3]

### 1 - Simple Knapsack Solver

In [5]:
solution = run_knapsack(values, weights, cap)

print("pour une résolution simple et sans tri, la valeur optimale est : \n",solution)

pour une résolution simple et sans tri, la valeur optimale est : 
 220


### 2 - Greedy Solver

In [5]:
solution = run_greedy(np.array(values), np.array(weights), cap)

print("\n pour une résolution en mode greedy et donc avec tri, la valeur optimale est : \n",solution)

valeurs =  [ 40  50 100  95  30]
poids =  [2 3 1 5 3]
la capacité maximale attendue est =  10 

Ajouter l'item < 100 , 1 > dans le sac
Ajouter l'item < 40 , 2 > dans le sac
Ajouter l'item < 95 , 5 > dans le sac

 position finale de chaque item dans le sac =  [1 0 1 1 0]

 leurs valeurs respectivent =  [ 40 100  95]

 leurs poids respectifs =  [2 1 5]

 poid du sac à la solution optimale =  8

 pour une résolution en mode greedy et donc avec tri, la valeur optimale est : 
 235


Nota : 1 on prend, 0 on ne prends pas !!!

### 3 - Récursivité sans mémorisation

In [137]:
datas = get_datas_in_an_other_format(values,weights)

solution = run_recursivity(0,datas,cap,{})

print("la solution avec récursivité et sans mémorsation : ",solution)

la solution avec récursivité et sans mémorsation :  (245, {4: 1, 3: 0, 2: 1, 1: 0, 0: 0})


### 4 - Récursivité avec memo

In [7]:
solution = run_recursivity_memo(0,datas,cap,{},{})

print("la solution avec récursivité est : ",solution)

la solution avec récursivité est :  (245, {4: 1, 3: 0, 2: 1, 1: 1, 0: 0})


### 5 - Dynamic Programmation (DPi)

In [8]:
solution = run_bottum_up(datas, cap)

print('la solution en DP(0) est = ', solution)

la solution en DP(0) est =  225


### 6 - Branch & Bound

In [9]:
kpbb = KPBB(cap, values, weights)
solution = kpbb.solve_()
print("la solution est en best first: ", solution)

la solution est en best first:  220


### 7 - Simulated Annealing  (Récuit Simulé)

In [10]:
Sol, Val = run_annealing(cap,values,weights)

print("Les items en 1/0 sont : ", str(Sol), " et, la valeur optimale (améliorante) est : ", str(Val))

Les items en 1/0 sont :  [0, 1, 1, 1, 0]  et, la valeur optimale (améliorante) est :  245


/home/ketsia/Bureau/bureau/TP_Opti_Combi_MULAPI_TITA_Ketsia/methods/knapsack_recuit_simule.py:53: RuntimeWarning: overflow encountered in exp
  acceptance_probability = min(np.exp(proposed_score-current_score), 1)


### 8 - Algorithme Génétique

In [155]:
datas = get_datas_in_an_other_format(values,weights)

print(datas)

params = {
    "vals" : values,
    "weis" : weights,
    "mutation_percentage": 0.5,
    "select_top": 3,
    "generation_size": 5,
    "fit_threshold": 0.5,
    "max_generations": 5,
    "max_weight": cap,
    "max_per_item": 2,
    "items": datas
}

sac = kp_genetic(params)

res, vo, p = sac.run_genetic()

print(res)
print("valeur opti : ",vo)
print("\n poids",p)

[(40, 2), (50, 3), (100, 1), (95, 5), (30, 3)]
Generation #1:	[1, 2, 0, 0, 2]	-1
Generation #2:	[0, 1, 0, 1, 0]	-1
Generation #3:	[1, 1, 0, 0, 2]	-1
Generation #4:	[0, 1, 0, 1, 0]	-1
Generation #5:	[1, 0, 0, 1, 0]	-1
[1, 1, 0, 1, 0]
valeur opti :  185

 poids 10


## Dataset Knapsack experience

#### Dans cette section on va s'intérresser à 2 datasets en raison de 2 par (dossier) échelle ou à la dimension (large_scale, low-dimensional) et pour l'ensemble de nos algorithmes.

- ce site nous a aidé à comprendre les fichiers :

http://artemisa.unicauca.edu.co/~johnyortega/instances_01_KP/

- et ce repository github à comparer les résultats :

https://github.com/Lolik-Bolik/Knapsack_problem_0-1

In [5]:
racine="datas/instances_01_KP"

# dans l'ordre indiqué 

# dataset 1 : 200 items, valeur optimale (2697), Capacité (997)
V1, W1, C1, VW1, n1 = get_knapsack_datas("knapPI_3_200_1000_1","large_scale",racine)

# dataset 2 : 6 items, valeur optimale (130), Capacité (80)
V2, W2, C2, VW2, n2 = get_knapsack_datas("f9_l-d_kp_5_80","low-dimensional",racine)

# On retient dataset 1 et dataset 2

### A) Simple solver avec dataset 1 & 2

In [41]:
print("DATASET 1 - SOLUTION = ",run_knapsack(V1, W1, C1))
print("DATASET 2 - SOLUTION = ",run_knapsack(V2, W2, C2))

DATASET 1 - SOLUTION =  1693
DATASET 2 - SOLUTION =  130


### B) Greedy solver avec dataset 1 & 2

In [5]:
print("\n\n\nDATASET 1 - SOLUTION : ",run_greedy(np.array(V1), np.array(W1), C1))

valeurs =  [ 585  194  426  606  348  516  521 1092  422  749  895  337  143  557
  945  915 1055  546  352  522  109  891 1001  459  222  767  194  698
  838  107  674  644  815  434  982  866  467 1094 1084  993  399  733
  533  231  782  528  172  800  974  717  238  974  956  820  245  519
 1095  894  629  296  299 1097  377  216  197 1008  819  639  342  807
  207  669  222  637  170 1031  198  826  700  587  745  872  367  613
 1072  181  995 1043  313  158  848  403  587  864 1023  636  129  824
  774  889  640  579  654  242  567  439  146  741  810  296  653  594
  291  166  824  924  830  308 1088  811  190  900  440  414  649  389
  296  501  965  566  778  789  670  933 1036  325  822  344  751  949
  223  213  531  479  608  461  685  165  953  586  742  786 1092  386
  825  989  386  124  912  591  959  991  763  190  188  281  279  314
  287  117  719  572  361  518  946  519  292  456  361  782  614  406
  986  301  630  485  949 1052  394  600  899  294  491  837  430 

In [39]:
print("\n\n\n DATASET 2 - SOLUTION : ",run_greedy(np.array(W2), np.array(W2), C2))

valeurs =  [15 20 17  8 31]
poids =  [15 20 17  8 31]
la capacité maximale attendue est =  80 

Ajouter l'item < 15 , 15 > dans le sac
Ajouter l'item < 20 , 20 > dans le sac
Ajouter l'item < 17 , 17 > dans le sac
Ajouter l'item < 8 , 8 > dans le sac

 position finale de chaque item dans le sac =  [1 1 1 1 0]

 leurs valeurs respectivent =  [15 20 17  8]

 leurs poids respectifs =  [15 20 17  8]

 poid du sac à la solution optimale =  60



 DATASET 2 - SOLUTION :  60


### C) Récursivité sans mémorisation avec dataset 1 & 2

In [ ]:
print("DATASET 1 - SOLUTION : ",run_recursivity(0,get_datas_in_an_other_format(V1,W1),C1,{}))

In [43]:
print("DATASET 2 - SOLUTION : ",run_recursivity(0,get_datas_in_an_other_format(V2,W2),C2,{}))

DATASET 2 - SOLUTION :  (130, {4: 0, 3: 1, 2: 1, 1: 1, 0: 1})


### D) Récursivité avec memo avec dataset 1 & 2

In [11]:
print("DATASET 1 - SOLUTION : ",run_recursivity_memo(0,get_datas_in_an_other_format(V1,W1),C1,{},{}))

DATASET 1 - SOLUTION :  (2696, {199: 1, 198: 1, 197: 0, 196: 0, 195: 0, 194: 1, 193: 0, 192: 0, 191: 0, 190: 0, 189: 0, 188: 0, 187: 0, 186: 0, 185: 1, 184: 0, 183: 1, 182: 0, 181: 0, 180: 0, 179: 0, 178: 1, 177: 0, 176: 1, 175: 0, 174: 0, 173: 0, 172: 0, 171: 0, 170: 0, 169: 1, 168: 0, 167: 1, 166: 1, 165: 1, 164: 1, 163: 1, 162: 0, 161: 0, 160: 0, 159: 0, 158: 0, 157: 1, 156: 0, 155: 0, 154: 0, 153: 0, 152: 0, 151: 0, 150: 0, 149: 0, 148: 0, 147: 1, 146: 0, 145: 0, 144: 0, 143: 0, 142: 0, 141: 0, 140: 0, 139: 0, 138: 0, 137: 0, 136: 0, 135: 0, 134: 0, 133: 0, 132: 0, 131: 0, 130: 0, 129: 0, 128: 0, 127: 0, 126: 0, 125: 0, 124: 0, 123: 0, 122: 0, 121: 0, 120: 1, 119: 0, 118: 0, 117: 0, 116: 0, 115: 0, 114: 0, 113: 1, 112: 0, 111: 0, 110: 0, 109: 0, 108: 0, 107: 0, 106: 1, 105: 0, 104: 0, 103: 1, 102: 0, 101: 0, 100: 0, 99: 0, 98: 0, 97: 0, 96: 1, 95: 0, 94: 0, 93: 0, 92: 0, 91: 0, 90: 0, 89: 1, 88: 0, 87: 0, 86: 0, 85: 1, 84: 0, 83: 0, 82: 0, 81: 0, 80: 0, 79: 0, 78: 0, 77: 0, 76: 0, 

In [5]:
print("DATASET 2 - SOLUTION : ",run_recursivity_memo(0,get_datas_in_an_other_format(V2,W2),C2,{},{}))

DATASET 2 - SOLUTION :  (130, {4: 0, 3: 1, 2: 1, 1: 1, 0: 1})


### E) Dynamic Programmation (DPi) avec Bottum_UP avec dataset 1 & 2

In [10]:
print("DATASET 1 - SOLUTION : ",run_bottum_up(get_datas_in_an_other_format(V1,W1), C1))

DATASET 1 - SOLUTION :  1791


In [6]:
print("DATASET 2 - SOLUTION : ",run_bottum_up(get_datas_in_an_other_format(V2,W2), C2))

DATASET 2 - SOLUTION :  109


### F) Branch & Bound avec dataset 1 & 2

In [10]:
print("la solution est en best first: ", KPBB(C1, V1, W1).solve_())

la solution est en best first:  1693


In [7]:
print("la solution est en best first: ", KPBB(C2, V2, W2).solve_())

la solution est en best first:  130


### G) Simulated Annealing  (Récuit Simulé) avec dataset 1 & 2

In [11]:
Sol, Val = run_annealing(C1,V1,W1)
print("DATASET 1 - solution : Les items en 1/0 sont : ", str(Sol), " et, la valeur optimale (améliorante) est : ", str(Val))

DATASET 1 - solution : Les items en 1/0 sont :  [0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  et, la valeur optimale (améliorante) est :  2394


In [8]:
Sol, Val = run_annealing(C2,V2,W2)
print("DATASET 2 - solution : Les items en 1/0 sont : ", str(Sol), " et, la valeur optimale (améliorante) est : ", str(Val))

DATASET 2 - solution : Les items en 1/0 sont :  [1, 1, 1, 1, 0]  et, la valeur optimale (améliorante) est :  130


/home/ec2-user/SageMaker/Opti_Combi/ZeTP/methods/knapsack_recuit_simule.py:53: RuntimeWarning: overflow encountered in exp
  acceptance_probability = min(np.exp(proposed_score-current_score), 1)


### H) Algorithme Génétique avec dataset 1 & 2

In [108]:
datas = get_datas_in_an_other_format(V1,W1)

print(datas)

params = {
    "vals" : V1,
    "weis" : W1,
    "mutation_percentage": 0.5,
    "select_top": 2,
    "generation_size": 2,
    "fit_threshold": 0.5,
    "max_generations": 5,
    "max_weight": C1,
    "max_per_item": 2,
    "items": datas
}

print("DATASET 2 - SOLUTION : ")

sac = kp_genetic(params)

res, vo, p = sac.run_genetic()

print(res)
print("la valeur optimale : ",vo)
print("le poids trouvé : ",p)

[(585, 485), (194, 94), (426, 326), (606, 506), (348, 248), (516, 416), (521, 421), (1092, 992), (422, 322), (749, 649), (895, 795), (337, 237), (143, 43), (557, 457), (945, 845), (915, 815), (1055, 955), (546, 446), (352, 252), (522, 422), (109, 9), (891, 791), (1001, 901), (459, 359), (222, 122), (767, 667), (194, 94), (698, 598), (838, 738), (107, 7), (674, 574), (644, 544), (815, 715), (434, 334), (982, 882), (866, 766), (467, 367), (1094, 994), (1084, 984), (993, 893), (399, 299), (733, 633), (533, 433), (231, 131), (782, 682), (528, 428), (172, 72), (800, 700), (974, 874), (717, 617), (238, 138), (974, 874), (956, 856), (820, 720), (245, 145), (519, 419), (1095, 995), (894, 794), (629, 529), (296, 196), (299, 199), (1097, 997), (377, 277), (216, 116), (197, 97), (1008, 908), (819, 719), (639, 539), (342, 242), (807, 707), (207, 107), (669, 569), (222, 122), (637, 537), (170, 70), (1031, 931), (198, 98), (826, 726), (700, 600), (587, 487), (745, 645), (872, 772), (367, 267), (613,

In [84]:
datas = get_datas_in_an_other_format(V2,W2)

print(datas)

params = {
    "vals" : V2,
    "weis" : W2,
    "mutation_percentage": 0.5,
    "select_top": 3,
    "generation_size": 5,
    "fit_threshold": 0.5,
    "max_generations": 5,
    "max_weight": C2,
    "max_per_item": 2,
    "items": datas
}

print("DATASET 2 - SOLUTION : \n")

sac = kp_genetic(params)

res, vo, p  = sac.run_genetic()

print(res)
print("la valeur optimale : ",vo)
print("le poids trouvé : ",p)

[(33, 15), (24, 20), (36, 17), (37, 8), (12, 31)]
DATASET 2 - SOLUTION : 

Generation #1:	[1, 1, 1, 2, 0]	-1
25
25
[0, 0, 1, 2, 0]
la valeur optimale :  110
le poids trouvé :  33


## Perspective : Recherche TABOU

Grace à ce Travail Pratique, j'ai pu découvrire une autre heuristique du nom de TABOU_SEARCH (Rehcerhce Tabou) :

Selon Jacques Teghem, la recherche Tabou, est une META HEURISTIQUE de recherche locale qui tente de remédier à l'inconvénient majeur d'une méthode de descente de rester bloquer dans un optimum local (donc il joue le même rôle que le recuit simulé), on se base sur le voisinage.

# Comparaison des résultats et Conclusion

_______________
NOM DATASET 1 : **KnapPI_3_200_1000_1** / VALEUR OPTIMALE ATTENDUE : **2697** / capacité attendue : **997** /  ITEMS : **200**
_________________



|   Algorithme            |Valeur opitmale trouvée|  Capacité  trouvée|Observation    |
|-------------------------|:-:                    |:-:                |:-:            |
|Simple Solver            |        1693           |   ####            |     ###       |
|Greedy Solver            |        2649           |   949             |     Bon       |
|Recursivité sans mémo    |                       |   ####            |     ###       |
|Récursivité avec mémo    |        2696           |   ####            |     ###       | 
|Dynamic Prog avec BottumU|        1791           |   ####            |     ###       |
|Branch & Bound           |        1693           |   +986            |    Pas mal    |
|Recuit Simulé            |        2394           |   ####            |     ###       |
|Genetique                |        ####           |   ####             |    Pas bon    |

___________________
NOM DATASET 2 : **f9_l-d_Kp_5_80** / VALEUR OPTIMALE ATTENDUE : **130** / capacité attendue : **80** / ITEMS : **5**
___________________

|   Algorithme            |Valeur opitmale trouvée|  Capacité  trouvée|Observation       |
|-------------------------|:-:                    |:-:                |:-:               |
|Simple Solver            |        130            |   60              |    Bon           |
|Greedy Solver            |        60             |   60              |    Pas mal       |
|Recursivité sans mémo    |        130            |   60              |    Bon           |
|Récursivité avec mémo    |        130            |   60              |    Bon           |  
|Dynamic Prog avec BottumU|        109            |   60              |    Bon           |
|Branch & Bound           |        130            |   60              |    Bon           |
|Recuit Simulé            |        130            |   60              |    Bon           |
|Genetique                |        110            |   33              |    Pas bon       |  

En définitive, ce travail consistait à mettre en place un certain nombre d'algorithme en Optimisation Combinatoire, en utilisant le problème de sac à dos (Knapsack).

Honnéteté scientifique oblige, j'ai été inspiré par certains reposit. github afin de pouvoir pallier aux difficultés ou, aux bugs rencontrés, d'autre part, les résultats ont été concluante pour le jeux de données de faibles instances, j'ose croire donc qu'il est possible de résoudre un problème combinatoire (1/0) en utilisant des méthodes sur la base d'un criètere, méthodes qui permettront d'atteindre un resultat.

certains algorithmes n'ont pas de valeurs car ils ont mis beaucoup de temps pour l'éxécution c'est le cas notamme de la récursivité sans mémorisation, d'autres par contre ont été plus rapide tel que greedy.